src/DSA.jl

```diff
@@ -1,174 +1,313 @@
 # This file is a part of JuliaFEM.
 # License is MIT: see https://github.com/JuliaFEM/Materials.jl/blob/master/LICENSE
 
-@with_kw mutable struct DSADriverState <: AbstractMaterialState
-    time::Float64 = zero(Float64)
-    strain::SymmetricTensor{2,3} = zero(SymmetricTensor{2,3,Float64})
+module DSAModule
+
+using LinearAlgebra, ForwardDiff, Tensors, NLsolve, Parameters
+
+import ..AbstractMaterial, ..AbstractMaterialState
+import ..Utilities: Symm2, Symm4, isotropic_elasticity_tensor, lame, debang
+import ..integrate_material!  # for method extension
+
+# parametrically polymorphic for any type representing ℝ
+export GenericDSA, GenericDSADriverState, GenericDSAParameterState, GenericDSAVariableState
+
+# specialization for Float64
+export DSA, DSADriverState, DSAParameterState, DSAVariableState
+
+@with_kw mutable struct GenericDSADriverState{T <: Real} <: AbstractMaterialState
+    time::T = zero(T)
+    strain::Symm2{T} = zero(Symm2{T})
+end
+
+# TODO: complete this docstring
+"""Parameter state for DSA (dynamic strain aging) material.
+
+This is similar to the Chaboche model, but with additional static recovery terms.
+
+`E`: Young's modulus
+`nu`: Poisson's ratio
+`R0`: initial yield strength
+`Kn`: plasticity multiplier divisor (drag stress)
+`nn`: plasticity multiplier exponent
+`C1`, `D1`: parameters governing behavior of backstress X1
+`C2`, `D2`: parameters governing behavior of backstress X2
+`Q`: shift parameter for yield strength evolution
+`b`: multiplier for yield strength evolution
+`w`: ???
+`P1`, `P2`: ???
+`m`: ???
+`m1`, `m2`: ???
+`M1`, `M2`: ???
+`ba`: ???
+`xi`: ???
+"""
+@with_kw struct GenericDSAParameterState{T <: Real} <: AbstractMaterialState
+    E::T = 0.0
+    nu::T = 0.0
+    R0::T = 0.0
+    Kn::T = 0.0
+    nn::T = 0.0
+    C1::T = 0.0
+    D1::T = 0.0
+    C2::T = 0.0
+    D2::T = 0.0
+    Q::T = 0.0
+    b::T = 0.0
+    w::T = 0.0
+    P1::T = 0.0
+    P2::T = 0.0
+    m::T = 0.0
+    m1::T = 0.0
+    m2::T = 0.0
+    M1::T = 0.0
+    M2::T = 0.0
+    ba::T = 0.0
+    xi::T = 0.0
+end
+
+# TODO: complete this docstring
```

Comment:
A model that captures dynamic (and static) strain aging (DSA) induced hardening. The related phenomena are:
- Portevin le Chatelier effect (Serrated yield, plastic instabilities, see https://doi.org/10.1016/j.commatsci.2011.05.039)
- Discontinuous yielding
- Inverse strain rate sensitivity
- Secondary hardening in low cycle fatigue (LCF) tests
These typically occur in certain temperature/strain rate regime, where the dislocations are pinned due to diffusion of solute atoms. In the most effective conditions, the speed of diffusion is comparable to the applied strain rate (speed of dislocations).